L58-python-async-api.md

```diff
@@ -0,0 +1,1673 @@
+
+    async def SayHello(self,
+                       request: helloworld_pb2_grpc.HelloRequest,
+                       context: grpc.aio.ServicerContext
+                ) -> helloworld_pb2_grpc.HelloReply:
+        await asyncio.sleep(1)
+        return helloworld_pb2.HelloReply(message="Hello, %s!" % request.name)
+
+server = grpc.aio.server()
+server.add_insecure_port(":50051")
+helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(), server)
+server.start()
+await server.wait_for_termination()
+```
+
+Client side:
+```Python
+async with grpc.aio.insecure_channel("localhost:50051") as channel:
+    stub = echo_pb2_grpc.EchoStub(channel)
+    response = await stub.Hi(echo_pb2.EchoRequest(message="ping"))
+
+    async for response in stub.StreamingHi(...):
+        process(response)
+```
+
+### New Streaming API - Reader / Writer API
+
+Existing streaming API has usability problem that its logic complexes user
+application. For client side, it requires the request iterator to be defined
+above the line of call, and the consumption of responses to be coded below the
+line of invocation. So when the application wants to send request based on
+received response, it would be quite challenging and usually involves
+multi-threading, which is not optimal in Python world.
+
+For server side, the same problem remains for non-trivial streaming servicer
+handlers. The `yield` syntax constraint the message sending should happen in the
+handler itself, otherwise the user needs to code a dedicated iterator class. For
+larger application, it would be a pain to create such iterator class for each
+streaming handler.
+
+euroelessar@ points out several great reasons for the new streaming API:
+* Enable `with` statement for streaming API, which allows clean resource management;
+* Easier to buffer outbound messages when peer pushes back (e.g. `yield` will
+  stop the execution of current coroutine, but `call.write` can be scheduled in
+  event loop);
+* Enable application to have both pending read and pending write in one
+  coroutine, and react on the first finisher (with iterator syntax this will be
+  quite hard to get it right).
+
+So, this gRFC introduces a new pattern of streaming API that reads/writes
+message to peer with explicit call.
+
+#### Snippet For Reader / Writer Streaming API
+
+```Python
+### Client side
+with stub.StreamingHi() as streaming_call:
+  request = echo_pb2.EchoRequest(message="ping")
+  await streaming_call.write(request)
+  response = await streaming_call.read()
+  while response:  # or response is not grpc.aio.EOF
+      process(response)
+      response = await streaming_call.read()
+
+### Server side
+class AsyncGreeter(helloworld_pb2_grpc.GreeterServicer):
+
+    async def StreamingHi(self,
+                          unused_request_iterator,
+                          context: grpc.aio.ServicerContext
+            ) -> None:
+        bootstrap_request = await context.read()
```

Comment:
thought: having access to context.read() might not be necessary as (IMHO) anext(request_iterator) does exactly the same thing, but it it might make sense to have servicer_context.read() for consistency as on the client side, one can read using call.read() (and there's no iterator access available on the client side?)